CONTRIBUTING.md

```diff
@@ -121,3 +122,120 @@ run it before submitting your PR. It runs `gofmt -s` (simplify) and `golint`.
+interaction with a human (such as this Collector).
+
+### Startup Error Handling
+
+Verify configuration during startup and fail fast if the configuration is invalid.
+This will bring the attention of a human to the problem as it is more typical for humans
+to notice problems when the process is starting as opposed to problems that may arise
+sometime (potentially long time) after process startup. Monitoring systems are likely
+to automatically flag processes that exit with failure during startup, making it
+easier to notice the problem. The Collector should print a reasonable log message to
+explain the problem and exit with a non-zero code. It is acceptable to crash the process
+during startup if there is no good way to exit cleanly but do your best to log and
+exit cleanly with a process exit code.
+
+### Do not Crash after Startup
+
+Do not crash or exit the Collector process after the startup sequence is finished.
+A running Collector typically contains data that is received but not yet exported further
+(e.g. is stored in the queues and other processors). Crashing or exiting the Collector
+process will result in losing this data since typically the receiver has
+already acknowledged the receipt for this data and the senders of the data will
+not send that data again. This recommendation may change in the future if we
+implement persistent queues in the Collector.
+
+### Bad Input Handling
+
+Do not crash on bad input in receivers or elsewhere in the pipeline.
+[Crash-only software](https://en.wikipedia.org/wiki/Crash-only_software)
+is valid in certain cases however this is not a correct approach for Collector (except
+during startup, see above). The reason is that many senders from which Collector
+receives data have built-in automatic retries of the _same_ data if no
+acknowledgment is received from the Collector. If you crash on bad input
+chances are high that after the Collector is restarted it will see the same
+data in the input and will crash again. This will likely result in infinite
+crashing loop if you have automatic retries in place.
+
+Typically bad input when detected in a receiver should be reported back to the
+sender. If it is elsewhere in the pipeline it may be too late to send a response
+to the sender (particularly in processors which are not synchronously processing
+data). In either case it is recommended to keep a metric that counts bad input data.
+
+### Error Handling and Retries
+
+Be rigorous in error handling. Don't ignore errors. Think carefully about each
+error and decide if it is a fatal problem or a transient problem that may go away
+when retried. Fatal errors should be logged or recorded in an internal metric to
+provide visibility to users of the Collector. For transient errors come up with a
+retrying strategy and implement it. Typically you will
+want to implement retries with some sort of exponential back-off strategy. For
```

Comment:
Typically we don't want exporters to have their own retries: we want that to be handled by processors that can implement generic policies of retry. Having their own retries requires set of custom observability. Preference here is to implement exporters using the available helpers.

Reply:
Think of a case when exporter first needs to `Connect` and then send data. The `Connect` is usually done when exporter is starting and will typically have its own logic. This is already what e.g. OpenCensus exporter does.
I am not sure this can be handled well by `queued_retry` processor. Let's think a bit more and discuss this additionally.

Reply:
Yeah, this is a good discussion to have.